cpp/include/raft/core/serialize.hpp

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/core/detail/mdspan_numpy_serializer.hpp>
+#include <raft/core/device_mdspan.hpp>
+#include <raft/core/handle.hpp>
+#include <raft/core/host_mdspan.hpp>
+
+#include <iostream>
+#include <vector>
+
+/**
+ * Collection of serialization functions for RAFT data types
+ */
+
+namespace raft {
+
+template <typename ElementType, typename Extents, typename LayoutPolicy, typename AccessorPolicy>
+inline void serialize_mdspan(
+  const raft::handle_t& handle,
+  std::ostream& os,
+  const raft::host_mdspan<ElementType, Extents, LayoutPolicy, AccessorPolicy>& obj)
+{
+  static_assert(std::is_same_v<LayoutPolicy, raft::layout_c_contiguous> ||
```

Comment:
Oh, now I see that here are the asserts I was missing from numpy_serializer. I think it would make sense to move them closer to the actual code which relies on these assumptions.